dispatch/dispatch.go

```diff
@@ -133,6 +138,134 @@ func (d *Dispatcher) run(it provider.AlertIterator) {
 	}
 }
 
+// AlertGroup represents how alerts alerts exist within an aggrGroup
+type AlertGroup struct {
+	Alerts   []*EnrichedAlert `json:"alerts"`
+	Labels   model.LabelSet   `json:"labels,omitempty"`
+	Receiver string           `json:"receiver,omitempty"`
+}
+
+// EnrichedAlert is the representation of an alert, with status info and
+// fingerprint.
+type EnrichedAlert struct {
+	*types.Alert
+	Status      types.AlertStatus
+	Fingerprint string
+	Receivers   *[]string
+}
+
+type AlertGroups []*AlertGroup
+
+func (ag AlertGroups) Swap(i, j int)      { ag[i], ag[j] = ag[j], ag[i] }
+func (ag AlertGroups) Less(i, j int) bool { return ag[i].Labels.Before(ag[j].Labels) }
+func (ag AlertGroups) Len() int           { return len(ag) }
+
+// Groups returns a slice of AlertGroups from the dispatcher's internal state.
+func (d *Dispatcher) Groups(matchers []*labels.Matcher, receiverFilter *regexp.Regexp, silenced, inhibited, active bool) AlertGroups {
```

Comment:
Passing all those filtering parameters, doing the filter in this function and having to pass the `setAlertStatus` function to the dispatcher feels a bit awkward. Same for the fact that `EnrichedAlert` contains information that doesn't really concern the `dispatch` package.

What about passing callback functions to `dispatch.Groups` instead?

```go
// EnrichedAlert is an alert with its associated receivers.
type EnrichedAlert struct {
        *types.Alert
        Receivers *[]string
}


func (d *Dispatcher) Groups(routeFilter func(*Route) bool, alertFilter func(*types.Alert) bool) AlertGroups {
        ...
        for route, ags := range d.aggrGroups {
                if !routeFilter(route) {
                        continue
                }
                ...
                for _, ag := range ags {
                        ...
                        for a := range alerts {
                                if !alertFilter(a) {
                                        continue
                                }
```